src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.Events {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class EventPublisherEventTypesAnalyzer : DiagnosticAnalyzer {
+
+		private const string EventAttributeFullName = "D2L.LP.Distributed.Events.Domain.EventAttribute";
+
+		private const string IEventNotifierFullName = "D2L.LP.Distributed.Events.Domain.IEventNotifier";
+		private const string IEventPublisherFullName = "D2L.LP.Distributed.Events.Domain.IEventPublisher";
+		private static IEnumerable<string> PublishMethodNames = ImmutableArray.Create( "Publish", "PublishMany" );
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.EventTypeMissingEventAttribute
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			INamedTypeSymbol eventAttributeType = compilation.GetTypeByMetadataName( EventAttributeFullName );
+			if( eventAttributeType.IsNullOrErrorType() ) {
+				return;
+			}
+
+			IImmutableSet<ISymbol> genericPublishMethods = Enumerable
+				.Concat(
+					GetGenericPublishMethods( compilation, IEventNotifierFullName ),
+					GetGenericPublishMethods( compilation, IEventPublisherFullName )
+				)
+				.ToImmutableHashSet();
+
+			context.RegisterSyntaxNodeAction(
+					ctxt => {
+						if( ctxt.Node is InvocationExpressionSyntax invocation ) {

```

Comment:
I don't think this guard is necessary. Pretty sure I checked this in Roslyn's source. I just assume the cast passes when I write analyzers.

Reply:
i.e. if you want to pull it out of the context for your analyze method this works:

```cs
context.RegisterSyntaxNode(
  ctx => AnalyzeMethodInvocation(
    ctx,
    ctx.Node,
    eventAttributeType,
    genericPublishMethods
  )
);
```

Reply:
- [ ] remove unnecessary guard

Reply:
Switched to a cast.  I'd rather an invalid cast exception then a null reference exception.

Reply:
Don't worry, its just that they couldn't get the API they really wanted in C#. The cast will never fail.

Reply:
But either way is good, ya hard casting it is probably what I'd do too on second thought